### PR TITLE
[Merged by Bors] - feat: add theorem about nodal on Subgroup

### PR DESCRIPTION
**Pull request by BoltonBailey**
A lemma about the vanishing polynomial.

Co-authored-by: Pratyush Mishra <prat@upenn.edu>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Polynomial/Degree/Definitions.lean [hunk 1]
```diff
@@ -888,6 +888,12 @@ theorem leadingCoeff_add_of_degree_lt (h : degree p < degree q) :
     coeff_add, zero_add]
 #align polynomial.leading_coeff_add_of_degree_lt Polynomial.leadingCoeff_add_of_degree_lt
 
+theorem leadingCoeff_add_of_degree_lt_left [Semiring R] (p q : R[X]) (h : degree q < degree p) :
+    leadingCoeff (p + q) = leadingCoeff p := by
+  rw [add_comm]
+  apply leadingCoeff_add_of_degree_lt
+  exact h
```
Comment (ocfnash) [inline Mathlib/Data/Polynomial/Degree/Definitions.lean:895]:
```suggestion
theorem leadingCoeff_add_of_degree_lt' (h : degree q < degree p) :
    leadingCoeff (p + q) = leadingCoeff p := by
  rw [add_comm]
  exact leadingCoeff_add_of_degree_lt h
```

File: Mathlib/Data/Polynomial/Degree/Definitions.lean [hunk 2]
```diff
@@ -1306,6 +1312,21 @@ theorem degree_sub_le_of_le {a b : WithBot ℕ} (hp : degree p ≤ a) (hq : degr
     degree (p - q) ≤ max a b :=
 (p.degree_sub_le q).trans <| max_le_max ‹_› ‹_›
 
+theorem leadingCoeff_sub_of_degree_lt [Ring R] (p q : R[X])
+    (h : Polynomial.degree q < Polynomial.degree p) : (p - q).leadingCoeff = p.leadingCoeff := by
+  rw [sub_eq_add_neg]
+  apply leadingCoeff_add_of_degree_lt_left
+  rw [degree_neg]
+  exact h
```
Comment (ocfnash) [inline Mathlib/Data/Polynomial/Degree/Definitions.lean:1320]:
```suggestion
theorem leadingCoeff_sub_of_degree_lt (h : Polynomial.degree q < Polynomial.degree p) :
    (p - q).leadingCoeff = p.leadingCoeff := by
  rw [sub_eq_add_neg]
  rw [← q.degree_neg] at h
  exact leadingCoeff_add_of_degree_lt' h
```
Also if we're going to have both the left and right version of `leadingCoeff_add_of_degree_lt` then we should do the same for the `sub` version.

File: Mathlib/Data/Polynomial/Degree/Definitions.lean [hunk 3]
```diff
@@ -1306,6 +1312,21 @@ theorem degree_sub_le_of_le {a b : WithBot ℕ} (hp : degree p ≤ a) (hq : degr
...
+
+theorem leadingCoeff_sub_of_degree_eq [Ring R] (p q : R[X]) (h : degree p = degree q)
+    (hlc : leadingCoeff p - leadingCoeff q ≠ 0) :
+    leadingCoeff (p - q) = leadingCoeff p - leadingCoeff q := by
+  rw [sub_eq_add_neg, leadingCoeff_add_of_degree_eq, leadingCoeff_neg, sub_eq_add_neg]
+  · rw [h, degree_neg]
+  · rw [leadingCoeff_neg, <-sub_eq_add_neg]
+    exact hlc
```
Comment (ocfnash) [inline Mathlib/Data/Polynomial/Degree/Definitions.lean:1328]: Spawning subgoals by rewriting with missing hypotheses makes proofs harder to follow. It's better to avoid this.
```suggestion
theorem leadingCoeff_sub_of_degree_eq (h : degree p = degree q)
    (hlc : leadingCoeff p ≠ leadingCoeff q) :
    leadingCoeff (p - q) = leadingCoeff p - leadingCoeff q := by
  replace h : degree p = degree (-q) := by rwa [q.degree_neg]
  replace hlc : leadingCoeff p + leadingCoeff (-q) ≠ 0 := by
    rwa [← sub_ne_zero, sub_eq_add_neg, ← q.leadingCoeff_neg] at hlc
  rw [sub_eq_add_neg, leadingCoeff_add_of_degree_eq h hlc, leadingCoeff_neg, sub_eq_add_neg] 
```

File: Mathlib/LinearAlgebra/Lagrange.lean [hunk 4]
```diff
@@ -620,6 +654,32 @@ theorem eval_interpolate_not_at_node' (hvs : Set.InjOn v s) (hs : s.Nonempty)
   simp only [mul_div_mul_left _ _ (eval_nodal_not_at_node hx), Pi.one_apply, mul_one]
 #align lagrange.eval_interpolate_not_at_node' Lagrange.eval_interpolate_not_at_node'
 
+/--
+The vanishing polynomial on a multiplicative subgroup is of the form X ^ n - 1
+-/
+theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup (Units F)) [Fintype G] :
+  nodal G.carrier.toFinset (fun (x : Fˣ) => (x : F)) = X ^ (Fintype.card G) - C 1 := by
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/Lagrange.lean:661]:
```suggestion
/-- The vanishing polynomial on a multiplicative subgroup is of the form X ^ n - 1. -/
theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup Fˣ) [Fintype G] :
  nodal (G : Set Fˣ).toFinset ((↑) : Fˣ → F) = X ^ (Fintype.card G) - 1 := by
```

Comment (ocfnash) [reply Mathlib/LinearAlgebra/Lagrange.lean:661]: Also, I prefer `Finite` and `Nat.card` over `Fintype` and `Fintype.card` (since I usually care only _that_ something is finite rather than _how_) but the right choice depends on what pattern has been followed in nearby parts of the library.

In this case we could do:
```lean
attribute [local instance] Fintype.ofFinite

/-- The vanishing polynomial on a multiplicative subgroup is of the form X ^ n - 1. -/
theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup Fˣ) [Finite G] :
  nodal (G : Set Fˣ).toFinset ((↑) : Fˣ → F) = X ^ (Nat.card G) - 1 := by
```
but I'll leave it to you to decide which is best here.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: You said this about #5738 too, both of these PRs come from the same project.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: All of this is meant for cryptographic applications, where the types are always finite and decidableEq. It's weird to me that there are like three different types of finiteness.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: I guess the point is that if you are doing a theorem, which has no data in it, we should use the weakest version of finiteness `Finite` so that every other version of finiteness can use it.

I feel sad that there seems to be a difference between `Nat.card` and `Fintype.card`, I am worried that I will have to prove lemmas later rewriting these.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: I don't really get the point of this local instance change. My understanding is it makes the lemma use the instance. If I am going to have `[Finite G]` but also use the `Fintype.ofFinite` instance, why not just use `Fintype`?

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: `G` has to be cast to a `Finset` anyway, and I think `Finset` is the `Fintype` kind of finiteness, so it feels like this is the right notion here.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: Maybe what you are really advocating for is "we should use `Nat.card` everywhere we are currently using `Fintype.card`

I feel dumb, when I change the lemma to use `Nat.card` like you suggest, I have to fix the bugs. How to I fix the bugs? By putting `Nat.card_eq_fintype_card` everywhere.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: Maybe the real culprit is the entire fact that `Finset` uses the `Fintype` finiteness and not the `Finite` finiteness?

I feel like I could make an argument for "Lagrange.lean is a file where I would expect more computable notions of finiteness"

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: Someone should write a docstring in `Mathlib.Data.Fintype.Card` about the difference between `Finite`/`Fintype`/`Fintype + DecidableEq`.

Comment (ocfnash) [reply Mathlib/LinearAlgebra/Lagrange.lean]: Given that you need a `Finset` for `nodal`, and `Finset` wants a `Fintype`, there's a good case to leave this as you wrote it, using the data-bearing constructive finiteness (i.e., `Fintype`) rather than the better `Prop`-valued `Finite`. So let's just leave this as you wrote it.

Since I've brought this up before, I'll try to explain a bit: once you're talking about finiteness of subsets of a type there are two axes of finiteness giving a total of four possibilities:

1. Constructive, unbundled: `Fintype` + `Set.Finite`
2. Constructive, bundled: `Finset`
3. Non-constructive, unbundled: `Finite`
4. Non-constructive, bundled: missing

Of course it's ridiculous that such a simple concept has been made so complicated. The reason is that originally 1 & 2 were created and there is an excellent library of results about them but eventually their constructive nature becomes problematic. So we created 3. This is much better  but not used as much yet. I try to push it because it's the future. Maybe some hero will one day swap out `Fintype` for `Finite` in `Finset` and so turn 2 into 4.

File: Mathlib/LinearAlgebra/Lagrange.lean [hunk 5]
```diff
@@ -620,6 +650,33 @@ theorem eval_interpolate_not_at_node' (hvs : Set.InjOn v s) (hs : s.Nonempty)
...
+    · rw [monic_X_pow]
+    · rw [degree_one, degree_pow, degree_X, nsmul_eq_mul, mul_one, Nat.cast_pos]
+      exact Fintype.card_pos
+  · intros i hi
+    have i_mem_G : i ∈ Set.toFinset G := hi
+    rw [eval_nodal_at_node i_mem_G]
+    simp at hi
+    rcases (CanLift.prf i hi : ∃ g : G, g.val = i) with ⟨g, g_eq⟩
+    rw [← g_eq, eval_sub, eval_pow, eval_X, eval_one, ← Units.val_pow_eq_pow_val,
+      ← Subgroup.coe_pow G, pow_card_eq_one, Subgroup.coe_one, Units.val_one, sub_self]
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/Lagrange.lean:678]: Here's a golf:
```suggestion
@[simp] theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup Fˣ) [Fintype G] :
  nodal (G : Set Fˣ).toFinset ((↑) : Fˣ → F) = X ^ (Fintype.card G) - 1 := by
  have h : degree (1 : F[X]) < degree ((X : F[X]) ^ Fintype.card G) := by simp [Fintype.card_pos]
  apply eq_of_degrees_le_of_eval_index_eq (v := ((↑) : Fˣ → F)) (G : Set Fˣ).toFinset
  · exact Set.injOn_of_injective Units.ext _
  · simp
  · rw [degree_sub_eq_left_of_degree_lt h, degree_nodal, Set.toFinset_card, degree_pow, degree_X,
      nsmul_eq_mul, mul_one, Nat.cast_inj]
    exact rfl
  · rw [nodal_monic, leadingCoeff_sub_of_degree_lt h, monic_X_pow]
  · intros i hi
    rw [eval_nodal_at_node hi]
    replace hi : i ∈ G := by simpa using hi
    obtain ⟨g, rfl⟩ : ∃ g : G, g.val = i := ⟨⟨i, hi⟩, rfl⟩
    simp [← Units.val_pow_eq_pow_val, ← Subgroup.coe_pow G]
```
Mostly this is just style nits but two small points are worth highlighting:

- There was a non-terminal `simp`
- Spawning subgoals via a `rw` with missing hypotheses makes proofs hard to follow, it's better to use `have` to establish the hypothesis and then feed it into the `rw`.

File: Mathlib/LinearAlgebra/Lagrange.lean [hunk 6]
```diff
@@ -101,6 +116,17 @@ theorem eq_of_degrees_lt_of_eval_index_eq (hvs : Set.InjOn v s) (degree_f_lt : f
   exact Submodule.sub_mem _ degree_f_lt degree_g_lt
 #align polynomial.eq_of_degrees_lt_of_eval_index_eq Polynomial.eq_of_degrees_lt_of_eval_index_eq
 
+theorem eq_of_degrees_le_of_eval_index_eq (hvs : Set.InjOn v s)
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/Lagrange.lean:119]:
```suggestion
theorem eq_of_degree_le_of_eval_index_eq (hvs : Set.InjOn v s)
```

**Review by ocfnash — APPROVED** [review]
Thanks, this is a useful result!

bors d+

File: Mathlib/LinearAlgebra/Lagrange.lean (hunk 5 above)
Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: Thanks I was wrong about the simp thing. For the `rw` thing, it strikes me that it tends to make the proof longer - is it something we are enforcing as a matter of course?

Comment (ocfnash) [reply Mathlib/LinearAlgebra/Lagrange.lean]: It is better to optimise for proof readability (and maintainability) than proof length.

You're right to raise this though: I don't think we mention this point about spawning subgoals with `rw` in our style guide. I might get it added to the new guide as we update it to Lean 4.

Comment (BoltonBailey) [reply Mathlib/LinearAlgebra/Lagrange.lean]: It surprises me that there is no non-terminal simp linter. Maybe this could be added as well?
